diff --git a/src/interface/editor_components/equalizer_response.cpp b/src/interface/editor_components/equalizer_response.cpp
index c2d005a..ad63740 100644
--- a/src/interface/editor_components/equalizer_response.cpp
+++ b/src/interface/editor_components/equalizer_response.cpp
@@ -181,9 +181,11 @@ void EqualizerResponse::drawResponse(OpenGlWrapper& open_gl, int index) {
                             band_filter_.getHighAmount()[index],
                             high_filter_.getHighAmount()[index]);
 
+  glEnable(GL_RASTERIZER_DISCARD);
   open_gl.context.extensions.glBeginTransformFeedback(GL_POINTS);
   glDrawArrays(GL_POINTS, 0, kResolution);
   open_gl.context.extensions.glEndTransformFeedback();
+  glDisable(GL_RASTERIZER_DISCARD);
 
   void* buffer = open_gl.context.extensions.glMapBufferRange(GL_TRANSFORM_FEEDBACK_BUFFER, 0,
                                                              kResolution * sizeof(float), GL_MAP_READ_BIT);
diff --git a/src/interface/editor_components/filter_response.cpp b/src/interface/editor_components/filter_response.cpp
index 0ede4c2..62a8b5d 100644
--- a/src/interface/editor_components/filter_response.cpp
+++ b/src/interface/editor_components/filter_response.cpp
@@ -203,7 +203,7 @@ void FilterResponse::init(OpenGlWrapper& open_gl) {
     shader->use();
     shaders_[i].position = getAttribute(open_gl, *shader, "position");
 
-    shaders_[i].mix = getUniform(open_gl, *shader, "mix");
+    shaders_[i].mix = getUniform(open_gl, *shader, "filter_mix");
     shaders_[i].midi_cutoff = getUniform(open_gl, *shader, "midi_cutoff");
     shaders_[i].resonance = getUniform(open_gl, *shader, "resonance");
     shaders_[i].drive = getUniform(open_gl, *shader, "drive");
@@ -615,9 +615,11 @@ void FilterResponse::drawFilterResponse(OpenGlWrapper& open_gl) {
 
 void FilterResponse::renderLineResponse(OpenGlWrapper& open_gl) {
   glEnable(GL_BLEND);
+  glEnable(GL_RASTERIZER_DISCARD);
   open_gl.context.extensions.glBeginTransformFeedback(GL_POINTS);
   glDrawArrays(GL_POINTS, 0, kResolution);
   open_gl.context.extensions.glEndTransformFeedback();
+  glDisable(GL_RASTERIZER_DISCARD);
 
   void* buffer = open_gl.context.extensions.glMapBufferRange(GL_TRANSFORM_FEEDBACK_BUFFER, 0,
                                                              kResolution * sizeof(float), GL_MAP_READ_BIT);
diff --git a/src/interface/editor_components/open_gl_component.h b/src/interface/editor_components/open_gl_component.h
index fe0fd89..5913731 100644
--- a/src/interface/editor_components/open_gl_component.h
+++ b/src/interface/editor_components/open_gl_component.h
@@ -74,7 +74,7 @@ class OpenGlComponent : public Component {
 
     static inline String translateFragmentShader(const String& code) {
     #if OPENGL_ES
-      return String("#version 300 es\n") + "out mediump vec4 fragColor;\n" +
+      return String("#version 300 es\nprecision mediump float;\n") + "out mediump vec4 fragColor;\n" +
              code.replace("varying", "in").replace("texture2D", "texture").replace("gl_FragColor", "fragColor");
     #else
       return OpenGLHelpers::translateFragmentShaderToV3(code);
@@ -83,7 +83,7 @@ class OpenGlComponent : public Component {
 
     static inline String translateVertexShader(const String& code) {
     #if OPENGL_ES
-      return String("#version 300 es\n") + code.replace("attribute", "in").replace("varying", "out");
+      return String("#version 300 es\nprecision mediump float;\n") + code.replace("attribute", "in").replace("varying", "out");
     #else
       return OpenGLHelpers::translateVertexShaderToV3(code);
     #endif
diff --git a/src/interface/editor_components/synth_preset_selector.cpp b/src/interface/editor_components/synth_preset_selector.cpp
index 64da1d5..01e7fb0 100644
--- a/src/interface/editor_components/synth_preset_selector.cpp
+++ b/src/interface/editor_components/synth_preset_selector.cpp
@@ -139,7 +139,8 @@ void SynthPresetSelector::newPresetSelected(File preset) {
     resetText();
   else {
     error = "There was an error open the preset. " + error;
-    AlertWindow::showNativeDialogBox("Error opening preset", error, false);
+    //AlertWindow::showNativeDialogBox("Error opening preset", error, false);
+    AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Error opening preset", error);
   }
 }
 
@@ -283,9 +284,8 @@ void SynthPresetSelector::importPreset() {
   SynthGuiInterface* parent = findParentComponentOfClass<SynthGuiInterface>();
   File active_file = parent->getSynth()->getActiveFile();
   FileChooser open_box("Open Preset", active_file, String("*.") + vital::kPresetExtension);
-  if (!open_box.browseForFileToOpen())
-    return;
-  
+
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
   File choice = open_box.getResult();
   if (!choice.exists())
     return;
@@ -294,10 +294,13 @@ void SynthPresetSelector::importPreset() {
   if (!parent->getSynth()->loadFromFile(choice, error)) {
     std::string name = ProjectInfo::projectName;
     error = "There was an error open the preset. " + error;
-    AlertWindow::showNativeDialogBox("Error opening preset", error, false);
+    //AlertWindow::showNativeDialogBox("Error opening preset", error, false);
+    AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Error opening preset", error);
   }
   else
     parent->externalPresetLoaded(choice);
+  };
+  open_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void SynthPresetSelector::exportPreset() {
@@ -308,16 +311,17 @@ void SynthPresetSelector::exportPreset() {
   SynthBase* synth = parent->getSynth();
   File active_file = synth->getActiveFile();
   FileChooser save_box("Export Preset", File(), String("*.") + vital::kPresetExtension);
-  if (!save_box.browseForFileToSave(true))
-    return;
-  
+
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {  
   synth->saveToFile(save_box.getResult().withFileExtension(vital::kPresetExtension));
   parent->externalPresetLoaded(synth->getActiveFile());
+  };
+  save_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 }
 
 void SynthPresetSelector::importBank() {
   FileChooser import_box("Import Bank", File(), String("*.") + vital::kBankExtension);
-  if (import_box.browseForFileToOpen()) {
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     File result = import_box.getResult();
     FileInputStream input_stream(result);
     if (input_stream.openedOk()) {
@@ -338,7 +342,8 @@ void SynthPresetSelector::importBank() {
     }
     else
       LoadSave::writeErrorLog("Opening file stream to bank failed!");
-  }
+  };
+  import_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void SynthPresetSelector::exportBank() {
@@ -348,8 +353,10 @@ void SynthPresetSelector::exportBank() {
 void SynthPresetSelector::loadTuningFile() {
   SynthGuiInterface* parent = findParentComponentOfClass<SynthGuiInterface>();
   FileChooser load_box("Load Tuning", File(), Tuning::allFileExtensions());
-  if (load_box.browseForFileToOpen())
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     parent->getSynth()->loadTuningFile(load_box.getResult());
+  };
+  load_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void SynthPresetSelector::clearTuning() {
@@ -404,12 +411,13 @@ void SynthPresetSelector::openSkinDesigner() {
 
 void SynthPresetSelector::loadSkin() {
   FileChooser open_box("Open Skin", File(), String("*.") + vital::kSkinExtension);
-  if (open_box.browseForFileToOpen()) {
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     File loaded = open_box.getResult();
     loaded.copyFileTo(LoadSave::getDefaultSkin());
     full_skin_->loadFromFile(loaded);
     repaintWithSkin();
-  }
+  };
+  open_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void SynthPresetSelector::clearSkin() {
diff --git a/src/interface/editor_sections/bank_exporter.cpp b/src/interface/editor_sections/bank_exporter.cpp
index ac1a476..ef68def 100644
--- a/src/interface/editor_sections/bank_exporter.cpp
+++ b/src/interface/editor_sections/bank_exporter.cpp
@@ -617,14 +617,15 @@ void BankExporter::exportBank() {
 
   File file = File::getCurrentWorkingDirectory().getChildFile(bank_name + "." + vital::kBankExtension);
   FileChooser export_box("Export Bank", file, String("*.") + vital::kBankExtension);
-  if (export_box.browseForFileToSave(true)) {
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     File destination = export_box.getResult().withFileExtension(vital::kBankExtension);
     if (destination.hasWriteAccess()) {
       FileOutputStream output_stream(destination);
       if (output_stream.openedOk())
         bank_zip.writeToStream(output_stream, nullptr);
     }
-  }
+  };
+  export_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 }
 
 void BankExporter::loadFiles() {
diff --git a/src/interface/editor_sections/distortion_section.cpp b/src/interface/editor_sections/distortion_section.cpp
index a7a3a3c..2e7a2a2 100644
--- a/src/interface/editor_sections/distortion_section.cpp
+++ b/src/interface/editor_sections/distortion_section.cpp
@@ -168,7 +168,7 @@ void DistortionFilterResponse::init(OpenGlWrapper& open_gl) {
   shader->use();
   response_shader_.position = getAttribute(open_gl, *shader, "position");
 
-  response_shader_.mix = getUniform(open_gl, *shader, "mix");
+  response_shader_.mix = getUniform(open_gl, *shader, "filter_mix");
   response_shader_.midi_cutoff = getUniform(open_gl, *shader, "midi_cutoff");
   response_shader_.resonance = getUniform(open_gl, *shader, "resonance");
   response_shader_.drive = getUniform(open_gl, *shader, "drive");
@@ -261,9 +261,11 @@ void DistortionFilterResponse::unbind(OpenGLContext& open_gl_context) {
 }
 
 void DistortionFilterResponse::renderLineResponse(OpenGlWrapper& open_gl) {
+  glEnable(GL_RASTERIZER_DISCARD);
   open_gl.context.extensions.glBeginTransformFeedback(GL_POINTS);
   glDrawArrays(GL_POINTS, 0, kResolution);
   open_gl.context.extensions.glEndTransformFeedback();
+  glDisable(GL_RASTERIZER_DISCARD);
 
   void* buffer = open_gl.context.extensions.glMapBufferRange(GL_TRANSFORM_FEEDBACK_BUFFER, 0,
                                                              kResolution * sizeof(float), GL_MAP_READ_BIT);
diff --git a/src/interface/editor_sections/download_section.cpp b/src/interface/editor_sections/download_section.cpp
index fccb2d5..df8bc96 100644
--- a/src/interface/editor_sections/download_section.cpp
+++ b/src/interface/editor_sections/download_section.cpp
@@ -316,7 +316,8 @@ void DownloadSection::triggerInstall() {
   if (!install_location_.exists() || !errors_file.exists() || !errors_file.hasWriteAccess()) {
     MessageManagerLock lock(Thread::getCurrentThread());
     String warning = "Can't create install directory. Select another destination";
-    AlertWindow::showNativeDialogBox("Can't Create Directory", warning, false);
+    //AlertWindow::showNativeDialogBox("Can't Create Directory", warning, false);
+    AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Can't Create Directory", warning);
     install_button_->setEnabled(true);
     cancel_button_->setEnabled(true);
     return;
@@ -383,6 +384,8 @@ void DownloadSection::cancelDownload() {
 }
 
 void DownloadSection::chooseInstallFolder() {
+  AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Not supported", "Choosing Install Directory not supported");
+  /*
   FileChooser open_box("Choose Install Directory", install_location_);
   if (open_box.browseForDirectory()) {
     File result = open_box.getResult();
@@ -399,7 +402,9 @@ void DownloadSection::chooseInstallFolder() {
     }
     else {
       String warning = "Can't create install directory. Select another destination";
-      AlertWindow::showNativeDialogBox("Invalid Directory", warning, false);
+      //AlertWindow::showNativeDialogBox("Invalid Directory", warning, false);
+      AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Invalid Directory", warning);
     }
   }
+  */
 }
diff --git a/src/interface/editor_sections/flanger_section.cpp b/src/interface/editor_sections/flanger_section.cpp
index f7911f9..9174359 100644
--- a/src/interface/editor_sections/flanger_section.cpp
+++ b/src/interface/editor_sections/flanger_section.cpp
@@ -84,7 +84,7 @@ void FlangerResponse::init(OpenGlWrapper& open_gl) {
   shader->use();
   response_shader_.position = getAttribute(open_gl, *shader, "position");
 
-  response_shader_.mix = getUniform(open_gl, *shader, "mix");
+  response_shader_.mix = getUniform(open_gl, *shader, "filter_mix");
   response_shader_.midi_cutoff = getUniform(open_gl, *shader, "midi_cutoff");
   response_shader_.resonance = getUniform(open_gl, *shader, "resonance");
   response_shader_.drive = getUniform(open_gl, *shader, "drive");
@@ -172,9 +172,11 @@ void FlangerResponse::unbind(OpenGLContext& open_gl_context) {
 void FlangerResponse::renderLineResponse(OpenGlWrapper& open_gl, int index) {
   static constexpr float kMaxMidi = 128.0f;
 
+  glEnable(GL_RASTERIZER_DISCARD);
   open_gl.context.extensions.glBeginTransformFeedback(GL_POINTS);
   glDrawArrays(GL_POINTS, 0, kResolution);
   open_gl.context.extensions.glEndTransformFeedback();
+  glDisable(GL_RASTERIZER_DISCARD);
 
   void* buffer = open_gl.context.extensions.glMapBufferRange(GL_TRANSFORM_FEEDBACK_BUFFER, 0,
                                                              kResolution * sizeof(float), GL_MAP_READ_BIT);
diff --git a/src/interface/editor_sections/lfo_section.cpp b/src/interface/editor_sections/lfo_section.cpp
index 170326e..8fdac12 100644
--- a/src/interface/editor_sections/lfo_section.cpp
+++ b/src/interface/editor_sections/lfo_section.cpp
@@ -360,18 +360,19 @@ void LfoSection::togglePaintMode(bool enabled, bool temporary_switch) {
 
 void LfoSection::importLfo() {
   FileChooser import_box("Import LFO", LoadSave::getUserLfoDirectory(), String("*.") + vital::kLfoExtension);
-  if (!import_box.browseForFileToOpen())
-    return;
 
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
   File choice = import_box.getResult();
   loadFile(choice.withFileExtension(vital::kLfoExtension));
+  };
+
+  import_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void LfoSection::exportLfo() {
   FileChooser export_box("Export LFO", LoadSave::getUserLfoDirectory(), String("*.") + vital::kLfoExtension);
-  if (!export_box.browseForFileToSave(true))
-    return;
 
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
   File choice = export_box.getResult();
   choice = choice.withFileExtension(vital::kLfoExtension);
   if (!choice.exists())
@@ -381,6 +382,8 @@ void LfoSection::exportLfo() {
   String name = choice.getFileNameWithoutExtension();
   editor_->getModel()->setName(name.toStdString());
   preset_selector_->setText(name);
+  };
+  export_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 }
 
 void LfoSection::fileLoaded() {
diff --git a/src/interface/editor_sections/master_controls_interface.cpp b/src/interface/editor_sections/master_controls_interface.cpp
index ec62969..95045fd 100644
--- a/src/interface/editor_sections/master_controls_interface.cpp
+++ b/src/interface/editor_sections/master_controls_interface.cpp
@@ -129,10 +129,13 @@ void TuningSelector::loadTuning(TuningStyle tuning) {
 void TuningSelector::loadTuningFile() {
   setCustomString("Custom");
   FileChooser load_box("Load Tuning", File(), Tuning::allFileExtensions());
-  if (load_box.browseForFileToOpen())
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     loadTuningFile(load_box.getResult());
 
+  // I assume this should be called only for successful case
   setCustomString(getTuningName().toStdString());
+  };
+  load_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void TuningSelector::loadTuningFile(const File& file) {
@@ -496,14 +499,15 @@ class DisplaySettings : public SynthSection {
         }
         else if (skin_->getValue() == skins_.size() + 1) {
           FileChooser open_box("Open Skin", File(), String("*.") + vital::kSkinExtension);
-          if (open_box.browseForFileToOpen()) {
+          std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
             File skin_file = open_box.getResult();
             skin_file.copyFileTo(LoadSave::getDefaultSkin());
             Skin skin;
 
             skin.loadFromFile(skin_file);
             loadSkin(skin);
-          }
+          };
+          open_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
         }
         else {
           Skin skin;
diff --git a/src/interface/editor_sections/modulation_matrix.cpp b/src/interface/editor_sections/modulation_matrix.cpp
index 740b0cf..0a103d8 100644
--- a/src/interface/editor_sections/modulation_matrix.cpp
+++ b/src/interface/editor_sections/modulation_matrix.cpp
@@ -1191,18 +1191,18 @@ void ModulationMatrix::togglePaintMode(bool enabled, bool temporary_switch) {
 
 void ModulationMatrix::importLfo() {
   FileChooser import_box("Import LFO", LoadSave::getUserLfoDirectory(), String("*.") + vital::kLfoExtension);
-  if (!import_box.browseForFileToOpen())
-    return;
 
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
   File choice = import_box.getResult();
   loadFile(choice.withFileExtension(vital::kLfoExtension));
+  };
+  import_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void ModulationMatrix::exportLfo() {
   FileChooser export_box("Export LFO", LoadSave::getUserLfoDirectory(), String("*.") + vital::kLfoExtension);
-  if (!export_box.browseForFileToSave(true))
-    return;
 
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
   File choice = export_box.getResult();
   choice = choice.withFileExtension(vital::kLfoExtension);
   if (!choice.exists())
@@ -1212,6 +1212,8 @@ void ModulationMatrix::exportLfo() {
   String name = choice.getFileNameWithoutExtension();
   map_editors_[selected_index_]->getModel()->setName(name.toStdString());
   preset_selector_->setText(name);
+  };
+  export_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 }
 
 void ModulationMatrix::fileLoaded() {
diff --git a/src/interface/editor_sections/phaser_section.cpp b/src/interface/editor_sections/phaser_section.cpp
index fc6ad0d..5594086 100644
--- a/src/interface/editor_sections/phaser_section.cpp
+++ b/src/interface/editor_sections/phaser_section.cpp
@@ -91,7 +91,7 @@ void PhaserResponse::init(OpenGlWrapper& open_gl) {
   shader->use();
   response_shader_.position = getAttribute(open_gl, *shader, "position");
 
-  response_shader_.mix = getUniform(open_gl, *shader, "mix");
+  response_shader_.mix = getUniform(open_gl, *shader, "filter_mix");
   response_shader_.midi_cutoff = getUniform(open_gl, *shader, "midi_cutoff");
   response_shader_.resonance = getUniform(open_gl, *shader, "resonance");
   response_shader_.db24 = getUniform(open_gl, *shader, "db24");
@@ -175,9 +175,11 @@ void PhaserResponse::unbind(OpenGLContext& open_gl_context) {
 }
 
 void PhaserResponse::renderLineResponse(OpenGlWrapper& open_gl) {
+  glEnable(GL_RASTERIZER_DISCARD);
   open_gl.context.extensions.glBeginTransformFeedback(GL_POINTS);
   glDrawArrays(GL_POINTS, 0, kResolution);
   open_gl.context.extensions.glEndTransformFeedback();
+  glDisable(GL_RASTERIZER_DISCARD);
 
   void* buffer = open_gl.context.extensions.glMapBufferRange(GL_TRANSFORM_FEEDBACK_BUFFER, 0,
                                                              kResolution * sizeof(float), GL_MAP_READ_BIT);
diff --git a/src/interface/editor_sections/popup_browser.cpp b/src/interface/editor_sections/popup_browser.cpp
index 13e827d..67ed851 100644
--- a/src/interface/editor_sections/popup_browser.cpp
+++ b/src/interface/editor_sections/popup_browser.cpp
@@ -517,6 +517,8 @@ void SelectionList::mouseDoubleClick(const MouseEvent& e) {
 }
 
 void SelectionList::addAdditionalFolder() {
+  AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Not supported", "Add Folder operation is not supported");
+  /*
   FileChooser open_box("Add Folder", File());
   if (open_box.browseForDirectory()) {
     File result = open_box.getResult();
@@ -545,6 +547,7 @@ void SelectionList::addAdditionalFolder() {
         showAddRootWarning();
     }
   }
+  */
 }
 
 void SelectionList::removeAdditionalFolder(const File& folder) {
diff --git a/src/interface/editor_sections/wavetable_edit_section.cpp b/src/interface/editor_sections/wavetable_edit_section.cpp
index a3a1ebe..8ec4b3a 100644
--- a/src/interface/editor_sections/wavetable_edit_section.cpp
+++ b/src/interface/editor_sections/wavetable_edit_section.cpp
@@ -576,20 +576,22 @@ void WavetableEditSection::saveAsWavetable() {
 
 void WavetableEditSection::importWavetable() {
   FileChooser open_box("Import Wavetable", File(), vital::kWavetableExtensionsList);
-  if (open_box.browseForFileToOpen()) {
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     if (!open_box.getResult().exists())
       return;
     loadFile(open_box.getResult());
-  }
+  };
+  open_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void WavetableEditSection::exportWavetable() {
   FileChooser save_box("Export Wavetable", File(), String("*.") + vital::kWavetableExtension);
-  if (save_box.browseForFileToSave(true)) {
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     json wavetable_data = wavetable_creator_->stateToJson();
     File file = save_box.getResult().withFileExtension(vital::kWavetableExtension);
     file.replaceWithText(wavetable_data.dump());
-  }
+  };
+  save_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 }
 
 void WavetableEditSection::exportToWav() {
@@ -597,9 +599,8 @@ void WavetableEditSection::exportToWav() {
   static constexpr int kNumWaveframes = 256;
   
   FileChooser save_box("Export to .wav File", File(), String("*.wav"));
-  if (!save_box.browseForFileToSave(true))
-    return;
 
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
   File file = save_box.getResult().withFileExtension("wav");
   if (!file.hasWriteAccess())
     return;
@@ -623,6 +624,8 @@ void WavetableEditSection::exportToWav() {
 
   writer = nullptr;
   file_stream.release();
+  };
+  save_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 }
 
 void WavetableEditSection::loadFile(const File& wavetable_file) {
diff --git a/src/interface/look_and_feel/shaders.cpp b/src/interface/look_and_feel/shaders.cpp
index 48fceb4..9626d3c 100644
--- a/src/interface/look_and_feel/shaders.cpp
+++ b/src/interface/look_and_feel/shaders.cpp
@@ -35,7 +35,7 @@
     "uniform " MEDIUMP " float midi_cutoff;\n" \
     "uniform " MEDIUMP " float resonance;\n" \
     "uniform " MEDIUMP " float drive;\n" \
-    "uniform " MEDIUMP " float mix;\n" \
+    "uniform " MEDIUMP " float filter_mix;\n" \
     "uniform " MEDIUMP " float db24;\n" \
     "uniform " MEDIUMP " float stage0;\n" \
     "uniform " MEDIUMP " float stage1;\n" \
@@ -158,7 +158,7 @@ namespace {
       "    gl_Position.x = gl_Position.x * scale.x;\n"
       "    gl_Position.y = gl_Position.y * scale.y;\n"
       "    gl_Position.z = 0.0;\n"
-      "    gl_Position.a = 1.0;\n"
+      "    gl_Position.w = 1.0;\n"
       "}\n";
 
   const char* kRotaryModulationVertexShader =
@@ -232,8 +232,9 @@ namespace {
       "    " MEDIUMP " vec2 response_res = complexDivide(two_stage, denominator);\n"
       "    " MEDIUMP " vec2 response = drive * response_res;\n"
       "    response = response + db24 * (complexMultiply(response_pre, response) - response);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kCombFilterResponseVertexShader =
@@ -254,7 +255,7 @@ namespace {
       "    " MEDIUMP " vec2 one_pole = stage0 * low_pass + stage1 * high_pass;\n"
       "    " MEDIUMP " vec2 high_pass2 = vec2(1.0, 0.0) - onePoleResponse(getCutoffRatio(position.x, stage3));\n"
       "    " MEDIUMP " vec2 filter_input = vec2(1.0 - 0.5 * abs(resonance), 0.0);\n"
-      "    " MEDIUMP " filter_input = complexMultiply(complexMultiply(filter_input, one_pole), high_pass2);\n"
+      "    " MEDIUMP " vec2 filter_input2 = complexMultiply(complexMultiply(filter_input, one_pole), high_pass2);\n"
       "    " MEDIUMP " vec2 denominator = vec2(1.0, 0.0) - complexMultiply(complexMultiply(tick, one_pole), high_pass2);\n"
       "    " MEDIUMP " float round_value = complexMultiply(one_pole, high_pass2).x * abs(resonance);\n"
       "    " MEDIUMP " vec2 denominator_round = vec2(1.0 - round_value, 0.0);\n"
@@ -264,9 +265,10 @@ namespace {
       "    denominator = max_step * max_step_denominator + (1.0 - max_step) * denominator;\n"
       "    " MEDIUMP " float denominator_round_step = 1.0 - max(max_step, step(ratio_diff, length(denominator)));\n"
       "    denominator = denominator_round_step * denominator_round + (1.0 - denominator_round_step) * denominator;\n"
-      "    " MEDIUMP " vec2 response = complexDivide(drive * filter_input, denominator);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    " MEDIUMP " vec2 response = complexDivide(drive * filter_input2, denominator);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kPositiveFlangeFilterResponseVertexShader =
@@ -299,8 +301,9 @@ namespace {
       "    " MEDIUMP " float denominator_round_step = 1.0 - max(max_step, step(ratio_diff, length(denominator)));\n"
       "    denominator = denominator_round_step * denominator_round + (1.0 - denominator_round_step) * denominator;\n"
       "    " MEDIUMP " vec2 response = filter_input * drive + complexMultiply(complexDivide(delay_input, denominator), delay);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kNegativeFlangeFilterResponseVertexShader =
@@ -330,8 +333,9 @@ namespace {
       "    " MEDIUMP " vec2 max_step_denominator = max_step_mult * denominator_round + (1.0 - max_step_mult) * denominator_round_down;\n"
       "    denominator = max_step * max_step_denominator + (1.0 - max_step) * denominator;\n"
       "    " MEDIUMP " vec2 response = filter_input * drive - complexMultiply(complexDivide(delay_input, denominator), delay);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kDigitalFilterResponseVertexShader =
@@ -352,8 +356,9 @@ namespace {
       "    " MEDIUMP " vec2 pre_response = complexDivide(numerator_pre, pre_denominator);\n"
       "    response = response + db24 * (complexMultiply(response, pre_response) - response);\n"
       "    response *= drive;\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kDiodeFilterResponseVertexShader =
@@ -378,9 +383,10 @@ namespace {
       "    " MEDIUMP " vec2 numerator = drive * chain;\n"
       "    " MEDIUMP " vec2 denominator = vec2(1.0, 0.0) + resonance * chain;\n"
       "    " MEDIUMP " vec2 response = complexDivide(numerator, denominator);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response = complexMultiply(response, high_pass_response);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kDirtyFilterResponseVertexShader =
@@ -405,8 +411,9 @@ namespace {
       "    " MEDIUMP " vec2 response_res = complexMultiply(two_stage, vec2(1.0, 0.0) + resonance_loop);\n"
       "    " MEDIUMP " vec2 response = drive * response_res;\n"
       "    response = response + db24 * (complexMultiply(response_pre, response) - response);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kFormantFilterResponseVertexShader =
@@ -444,8 +451,9 @@ namespace {
       "    " MEDIUMP " vec2 numerator3 = -low[3] * vec2(1.0, 0.0) + band[3] * vec2(0.0, g3) + high[3] * g3_sqr;\n"
       "    " MEDIUMP " vec2 response3 = complexDivide(numerator3, denominator3);\n"
       "    " MEDIUMP " vec2 response = response0 + response1 + response2 + response3;\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kLadderFilterResponseVertexShader =
@@ -465,10 +473,11 @@ namespace {
       "                              vec2(stage4, 0.0));\n"
       "    " MEDIUMP " vec2 denominator = four_pole_invert + vec2(resonance, 0.0);\n"
       "    " MEDIUMP " vec2 response = complexDivide(numerator, denominator);\n"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
-  
+
   const char* kPhaserFilterResponseVertexShader =
       "in " MEDIUMP " vec2 position;\n"
       "out " MEDIUMP " float response_out;\n"
@@ -491,8 +500,9 @@ namespace {
       "    " MEDIUMP " vec2 denominator = vec2(1.0, 0.0) - invert_mult * resonance * feedback_chain;\n"
       "    " MEDIUMP " vec2 phase_response = complexDivide(chain, denominator);\n"
       "    " MEDIUMP " vec2 response = vec2(0.5, 0.0) + 0.5 * invert_mult * phase_response;"
-      "    response = mix * response + vec2(1.0 - mix, 0.0);\n"
+      "    response = filter_mix * response + vec2(1.0 - filter_mix, 0.0);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kEqFilterResponseVertexShader =
@@ -529,6 +539,7 @@ namespace {
       "    " MEDIUMP " vec2 denominator = complexMultiply(denominator0, complexMultiply(denominator1, denominator2));\n"
       "    " MEDIUMP " vec2 response = complexDivide(numerator, denominator);\n"
       "    response_out = getYForResponse(response);\n"
+      "    gl_Position = vec4(0.0, 0.0, 0.0, 1.0);\n"
       "}\n";
 
   const char* kColorFragmentShader =
@@ -826,9 +837,9 @@ namespace {
       "uniform " MEDIUMP " float line_width;\n"
       "uniform " MEDIUMP " float boost;\n"
       "varying " MEDIUMP " float depth_out;\n"
-      "varying " MEDIUMP " float distance;\n"
+      "varying " MEDIUMP " float fill_distance;\n"
       "void main() {\n"
-      "    " MEDIUMP " vec4 color = color_to * distance + color_from * (1.0 - distance);\n"
+      "    " MEDIUMP " vec4 color = color_to * fill_distance + color_from * (1.0 - fill_distance);\n"
       "    " MEDIUMP " float dist_from_edge = min(depth_out, 1.0 - depth_out);\n"
       "    " MEDIUMP " float mult = 1.0 + boost * max(dist_from_edge - 2.0 / line_width, 0.0);\n"
       "    " MEDIUMP " vec4 result = mult * color;\n"
@@ -855,9 +866,9 @@ namespace {
       "uniform " MEDIUMP " vec4 color_from;\n"
       "uniform " MEDIUMP " vec4 color_to;\n"
       "varying " MEDIUMP " float boost;\n"
-      "varying " MEDIUMP " float distance;\n"
+      "varying " MEDIUMP " float fill_distance;\n"
       "void main() {\n"
-      "    " MEDIUMP " float delta = abs(distance);\n"
+      "    " MEDIUMP " float delta = abs(fill_distance);\n"
       "    " MEDIUMP " vec4 base_color = color_to * delta + color_from * (1.0 - delta);\n"
       "    gl_FragColor = base_color;\n"
       "    gl_FragColor.a = (boost + 1.0) * base_color.a;\n"
@@ -882,17 +893,17 @@ namespace {
       "uniform " MEDIUMP " vec2 scale;\n"
       "uniform " MEDIUMP " float center_position;\n"
       "uniform " MEDIUMP " float boost_amount;\n"
-      "out " MEDIUMP " float distance;\n"
+      "out " MEDIUMP " float fill_distance;\n"
       "out " MEDIUMP " float boost;\n"
       "\n"
       "void main() {\n"
-      "    distance = (position.y - center_position) / (1.0 - center_position);\n"
+      "    fill_distance = (position.y - center_position) / (1.0 - center_position);\n"
       "    boost = boost_amount * position.z;\n"
       "    gl_Position = position;\n"
       "    gl_Position.x = gl_Position.x * scale.x;\n"
       "    gl_Position.y = gl_Position.y * scale.y;\n"
       "    gl_Position.z = 0.0;\n"
-      "    gl_Position.a = 1.0;\n"
+      "    gl_Position.w = 1.0;\n"
       "}\n";
 
   const char* kBarFragmentShader =
@@ -950,7 +961,8 @@ namespace {
 
   inline String translateFragmentShader(const String& code) {
   #if OPENGL_ES
-    return String("#version 300 es\n") + "out mediump vec4 fragColor;\n" +
+    // FIXME: (Android) should this be highp? Did not improve visible scale so far...
+    return String("#version 300 es\nprecision mediump float;\n") + "out mediump vec4 fragColor;\n" +
       code.replace("varying", "in").replace("texture2D", "texture").replace("gl_FragColor", "fragColor");
   #else
     return OpenGLHelpers::translateFragmentShaderToV3(code);
@@ -959,7 +971,8 @@ namespace {
 
   inline String translateVertexShader(const String& code) {
   #if OPENGL_ES
-    return String("#version 300 es\n") + code.replace("attribute", "in").replace("varying", "out");
+    // FIXME: (Android) should this be highp? Did not improve visible scale so far...
+    return String("#version 300 es\nprecision mediump float;\n") + code.replace("attribute", "in").replace("varying", "out");
   #else
     return OpenGLHelpers::translateVertexShaderToV3(code);
   #endif
diff --git a/src/interface/look_and_feel/skin.cpp b/src/interface/look_and_feel/skin.cpp
index 13b1255..2bbdba7 100644
--- a/src/interface/look_and_feel/skin.cpp
+++ b/src/interface/look_and_feel/skin.cpp
@@ -641,20 +641,24 @@ class SkinColorPicker : public Component, public Button::Listener, public Slider
     void buttonClicked(Button* clicked_button) override {
       if (clicked_button == &load_button_) {
         FileChooser open_box("Open Skin", File(), String("*.") + vital::kSkinExtension);
-        if (open_box.browseForFileToOpen()) {
+        std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
           if (!skin_->loadFromFile(open_box.getResult())) {
-            AlertWindow::showNativeDialogBox("Error opening skin", "Skin file is corrupted and won't load.", false);
+            //AlertWindow::showNativeDialogBox("Error opening skin", "Skin file is corrupted and won't load.", false);
+            std::cout << "Error opening skin: Skin file is corrupted and won't load.";
             return;
           }
           setSliderValues();
           repaintWithSettings();
-        }
+        };
+        open_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
         return;
       }
       if (clicked_button == &save_button_) {
         FileChooser save_box("Save Skin", File(), String("*.") + vital::kSkinExtension);
-        if (save_box.browseForFileToSave(true))
+        std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
           skin_->saveToFile(save_box.getResult().withFileExtension(vital::kSkinExtension));
+        };
+        save_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
 
         return;
       }
diff --git a/src/interface/wavetable/overlays/file_source_overlay.cpp b/src/interface/wavetable/overlays/file_source_overlay.cpp
index 65e44a0..c577d98 100644
--- a/src/interface/wavetable/overlays/file_source_overlay.cpp
+++ b/src/interface/wavetable/overlays/file_source_overlay.cpp
@@ -443,8 +443,10 @@ void FileSourceOverlay::loadFile(const File& file) {
 
 void FileSourceOverlay::loadFilePressed() {
   FileChooser load_file("Load Audio File", File::getSpecialLocation(File::userHomeDirectory), String("*.wav"));
-  if (load_file.browseForFileToOpen())
+  std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
     loadFile(load_file.getResult());
+  };
+  load_file.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
 }
 
 void FileSourceOverlay::buttonClicked(Button* clicked_button) {
diff --git a/src/plugin/synth_plugin.cpp b/src/plugin/synth_plugin.cpp
index f215edd..e1127a6 100644
--- a/src/plugin/synth_plugin.cpp
+++ b/src/plugin/synth_plugin.cpp
@@ -210,7 +210,7 @@ void SynthPlugin::setStateInformation(const void* data, int size_in_bytes) {
   }
   catch (const json::exception& e) {
     std::string error = "There was an error open the preset. Preset file is corrupted.";
-    AlertWindow::showNativeDialogBox("Error opening preset", error, false);
+    AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Error opening preset", error);
   }
   pauseProcessing(false);
 
diff --git a/src/standalone/main.cpp b/src/standalone/main.cpp
index f8535fa..155c6eb 100644
--- a/src/standalone/main.cpp
+++ b/src/standalone/main.cpp
@@ -96,7 +96,8 @@ class SynthApplication : public JUCEApplication {
           if (!Startup::isComputerCompatible()) {
             String error = String(ProjectInfo::projectName) +
                            " requires SSE2, NEON or AVX2 compatible processor. Exiting.";
-            AlertWindow::showNativeDialogBox("Computer not supported", error, false);
+            //AlertWindow::showNativeDialogBox("Computer not supported", error, false);
+            std::cout << "Computer not supported";
             quit();
           }
 
@@ -185,8 +186,10 @@ class SynthApplication : public JUCEApplication {
           else if (info.commandID == kSaveAs) {
             File active_file = editor_->getActiveFile();
             FileChooser save_box("Export Preset", File(), String("*.") + vital::kPresetExtension);
-            if (save_box.browseForFileToSave(true))
+            std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
               editor_->saveToFile(save_box.getResult().withFileExtension(vital::kPresetExtension));
+            };
+            save_box.launchAsync(FileBrowserComponent::FileChooserFlags::saveMode, callback);
             grabKeyboardFocus();
             editor_->setFocus();
             return true;
@@ -194,9 +197,8 @@ class SynthApplication : public JUCEApplication {
           else if (info.commandID == kOpen) {
             File active_file = editor_->getActiveFile();
             FileChooser open_box("Open Preset", active_file, String("*.") + vital::kPresetExtension);
-            if (!open_box.browseForFileToOpen())
-              return true;
             
+            std::function<void(const FileChooser &)> callback = [&](const FileChooser &fc) {
             File choice = open_box.getResult();
             if (!choice.exists())
               return true;
@@ -204,12 +206,14 @@ class SynthApplication : public JUCEApplication {
             std::string error;
             if (!editor_->loadFromFile(choice, error)) {
               error = "There was an error open the preset. " + error;
-              AlertWindow::showNativeDialogBox("Error opening preset", error, false);
+              AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Error opening preset", error);
             }
             else
               editor_->externalPresetLoaded(choice);
             grabKeyboardFocus();
             editor_->setFocus();
+            };
+            open_box.launchAsync(FileBrowserComponent::FileChooserFlags::openMode, callback);
             return true;
           }
           else if (info.commandID == kToggleVideo)
diff --git a/src/synthesis/framework/common.h b/src/synthesis/framework/common.h
index f66ba88..e4b40da 100644
--- a/src/synthesis/framework/common.h
+++ b/src/synthesis/framework/common.h
@@ -18,6 +18,9 @@
 
 #include "JuceHeader.h"
 
+namespace juce { namespace gl { } }
+using namespace juce::gl; // It is required for JUCE 6.1.0 or later (see BREAKING-CHANGES.txt).
+
 // Debugging.
 #if DEBUG
 #include <cassert>
